aws/ec2.go

```diff
@@ -0,0 +1,225 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	awssession "github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	defaultTimeOut = 20
+	defaultRetries = 3
+)
+
+func convertError(err error) string {
+	if awsErr, ok := err.(awserr.Error); ok {
+		return fmt.Sprintf("%s: %s", awsErr.Code(), awsErr.Message())
```

Comment:
What does an AWS error normally stringify as?

Reply:
IIUC `Code`, `Message` and `OrigError`.

Reply:
So, so we really need this formatting function?  Would we get something reasonable with `fmt.Errorf("... %w", err)`?  Using `%w` has the effect that the original error is wrapped and preserved so the caller can inspect it if that ever becomes needed.

Reply:
yeah, I think this is needed. Based on [the documentation](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/handling-errors.html), `awserr` string-ifies Code, Message and OrigError when available.

To understand this better, I wrote a stub --
```
...
    interfaceID := "abcdef"
    srcDstCheckValue := true
    networkInterfaceAttribIn := &ec2.ModifyNetworkInterfaceAttributeInput{
        NetworkInterfaceId: aws.String(interfaceID),
        SourceDestCheck: &ec2.AttributeBooleanValue{
            Value: aws.Bool(srcDstCheckValue),
        },
    }

    networkInterfaceAttribOut, err := ec2Svc.ModifyNetworkInterfaceAttribute(networkInterfaceAttribIn)
    if err != nil {
        if aerr, ok := err.(awserr.Error); ok {
            fmt.Printf("err: %v\n", err)
            fmt.Printf("awserr: %s: %s\n", aerr.Code(), aerr.Message())
            return
...
```

here is the output for the stub --
```
err: InvalidNetworkInterfaceId.Malformed: Invalid id: "abcdef" (expecting "eni-...")
        status code: 400, request id: 57e804e8-0655-42bc-b26f-2d955eb7566f
awserr: InvalidNetworkInterfaceId.Malformed: Invalid id: "abcdef" (expecting "eni-...")
```

Notice the `OrigErr()` part of the output in the second line above. `OrigErr()` is not always set. But when set, shows _finer_ details which may not really useful to address the issue. When `OrigErr()` is not set (e.g. when AWS_REGION is not set), the output is identical for both Printf() in the stub. `convertError()` makes sure that the output always appears in the same format.

> Using %w has the effect that the original error is wrapped and preserved so the caller can inspect it if that ever becomes needed.

Regarding using `%w`, I am missing the point here. In general it's a good idea. However in this case, the caller is fixed. Had this code been part of a common API with several possible consumers, I do see the usefulness of error wrapping there. IMO, if this code were to ever change for such a use-case, we can adjust this code accordingly then.

For the current proposed change though, I don't see any disadvantage with the current format. Happy to be corrected if I missed any use-case or scenario.